### PR TITLE
Update biobambam - Clean messy build

### PR DESCRIPTION
**Pull request by matthdsm**
----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

For members of the Bioconda project, the following command is also available:

<table>
  <tr>
    <td><code>@BiocondaBot please merge</code></td>
    <td>Upload built packages/containers and merge a PR. <br />Someone must approve a PR first! <br />This reduces CI build time by reusing built artifacts.</td>
  </tr>
</table>

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

### CODE REVIEW

**Comment by matthdsm** [comment]
@BiocondaBot please add label

**Comment by jmarshall** [comment]
Thanks — that this compiled successfully confirms that these `exports` were unnecessary and serving only to confuse matters. I would also remove the `mkdir` as it is unnecessary (but at least that is not actively mysterious, as the exports were).

The source tarball contains the proper generated _configure_ files, so all the fiddling with `autoreconf` was also a red herring fixed by the eventual addition of `compiler('cxx')`. So you should also remove the `autoconf`/`automake`/`libtool` build requirements.

The `Needed DSO …` part of the build log shows that there is still some minor oddness about the library requirements here, but that is unchanged from previously and the executables in fact work — so I don't suggest any change is necessary for that.

The previous recipe just unpacked a precompiled binary tarball, but since your PR #24972 it builds bespoke executables for conda from the source tarball (yay!). I suspect this means that `binary_relocation: false` is no longer needed or desirable and should be removed, but that would need advice from a more expert maintainer.

**Comment by matthdsm** [comment]
Thanks for all the tips!
As I stated earlier, I'm not really at home in all the compiler and library stuff, so this is a great learning experience too.

I've made the requested modifications. Feel free to review after the checks passed.

Matthias

**Comment by matthdsm** [comment]
@gt1, I think we're going to need your help once again.
In PR #27336 I've updated libmaus2 to include snappy and io_lib. This seems to have worked, although I'm not really sure how to confirm this.

Now during the build of biobambam, we're still getting a notice io_lib is not available, while it most definitely is.
Are there any flags we're missing in your opinion? 

Thanks
Matthias

**Comment by gt1** [comment]
Hi Matthias,

I think this time I would really need to see the content of config.log to be able to tell.

German

**Comment by matthdsm** [comment]
@bioconda/build-system, @bioconda/core,
Is there anyone who can help us get some logs from the build? Is that even possible with CircleCI?

Thanks 
Matthias

**Comment by dpryan79** [comment]
@matthdsm The easiest thing to do is `cat` them after running ./configure. Then they're in the output that you can see. I can rerun a job and enable ssh so I can login, but since everything is run in docker containers that are cleaned up that ends up not being so helpful. I usually just rerun things locally and disable cleaning up the docker containers if this is really needed. Just let me know if you need me to do that.

**Comment by dpryan79** [comment]
Having said that, at least it's building and passing tests now, which is a good step in the right direction.

**Comment by matthdsm** [comment]
Thanks,
I'll start with the `cat`trick and see how we proceed from there on.

Matthias

**Comment by matthdsm** [comment]
@gt1, hope this is what you're looking for!

[config.log.txt](https://github.com/bioconda/bioconda-recipes/files/6139747/config.log.txt)

**Comment by matthdsm** [comment]
xref #27366, where I'm catting the config.log for the libmaus build

**Comment by matthdsm** [comment]
As far as I can decipher, snappy is fine, but libmaus wasn't properly configured with io_lib, so I think we'll have to debug there.

**Comment by dpryan79** [comment]
@bioconda-bot please fetch artifacts

**Comment by gt1** [comment]
I have just discovered an issue with io_lib detection in libmaus2. I'll provide a fix soon.

**Comment by matthdsm** [comment]
Thanks, I'll await your fix then!

Matthias

**Comment by gt1** [comment]
The current git version of io_lib causes a crash with libmaus2/biobambam2, so I have asked James Bonfield (io_lib author), whether this can be fixed. io_lib contains some unreleased additional features for libmaus2 (interface for CRAM encoding profiles), but apparently it got broken somewhere on the way. Ideally a fixed version of io_lib could reach release status, so this could be used. I'll see whether this can be achieved in a timely manner.

**Comment by matthdsm** [comment]
Great, thanks for following up!
If and when a new version op io_lib is released, I'll update the recipe.

M

**Comment by matthdsm** [comment]
Updating libmaus2 to latest version, with all fixes applied in #27366

File: recipes/biobambam/build.sh [hunk 1]
```diff
@@ -1,12 +1,8 @@
...
+./configure --prefix=${PREFIX}
 
-mkdir -p $PREFIX/bin
-export CPP_INCLUDE_PATH=${PREFIX}/include
-export CPLUS_INCLUDE_PATH=${PREFIX}/include
-export CXX_INCLUDE_PATH=${PREFIX}/include
-export LD_LIBRARY_PATH=${PREFIX}/lib
-export LIBRARY_PATH=${PREFIX}/lib
+#debug
+cat config.log
```
Comment (tomiles) [inline recipes/biobambam/build.sh:6]: Don't think we need to keep these lines?

**Review by tomiles — CHANGES_REQUESTED** [review]

File: recipes/biobambam/build.sh (hunk 1 above)
Comment (matthdsm) [reply recipes/biobambam/build.sh]: removed

**Review by dpryan79 — APPROVED** [review]

**Comment by dpryan79** [comment]
@bioconda-bot please merge

**Comment by matthdsm** [comment]
Thanks everyone!

**Comment by gt1** [comment]
Hi Matthias,

not sure what you are referring to, but if it is the /experimental/ ::experimental:: patches: these are are only necessary because the libstdc++ shipped with gcc7 does not have proper support for c++17. They are no longer necessary once you upgrade to gcc8 (or probably any other C++ environment fully supporting c++17).

Please let me know, if you were talking about something else.

Best,
German

**Comment by matthdsm** [comment]
Hi German.

I'm talking about this line
https://github.com/bioconda/bioconda-recipes/blob/f70760f5fd49dd086b3880bc6a949053967ab4fa/recipes/libmaus2/patch2#L8-L10

I added the include for `<experimental/filesystem>` instead of replacing the regular `<filesystem>`. This makes me think it's perhaps a bug in the source and not some artefact from building with an older gcc.

Matthias

**Comment by dpryan79** [comment]
FYI, I'm rolling out packages built with gcc9 starting this week.

**Comment by gt1** [comment]
Hi Matthias,

the file #includes another file #including <filesystem> which should be sufficient to get everything in the std::filesystem namespace.

Best,
German

**Comment by matthdsm** [comment]
Thanks for clarifying!

**Pull request merged**